drivers: ieee802154: rf2xx: Add EUI-64 from DT and iface no auto start

nandojve opened the pull request:
Add local-mac-address on DT and enable it on rf2xx driver. If user define local-mac-address this value will be used as default mac address. Otherwise driver automatically add a random mac address.

On application level user can change default mac address using net_mgmt command with NET_REQUEST_IEEE802154_SET_EXT_ADDR parameter defined on include/net/ieee802154_mgmt.h header.

~Refactor get_mac to rf2xx_get_mac and turn it as a weak method. This allows any user to create their own set mac address. This is important because RF2xx is a transceiver over SPI and may need mac address be defined when system is booting.~
Add a Kconfig option that sets NET_IF_NO_AUTO_START at driver init. This way user can control when net_if will be up with proper configuration.

Fixes #23193

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: ieee802154: rf2xx: Add local-mac-address

Add local-mac-address on DT and enable it on rf2xx driver. If user
define local-mac-address this value will be used as default mac address.
Otherwise driver automatically add a random mac address.

On application level user can change default mac address using net_mgmt
command with NET_REQUEST_IEEE802154_SET_EXT_ADDR parameter defined on
include/net/ieee802154_mgmt.h header.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -821,59 +824,70 @@ static struct ieee802154_radio_api rf2xx_radio_api = {
+#define DRV_INST_SPI_DEV_CS_GPIOS_FLAGS(n)                             \
+	UTIL_AND(DT_INST_SPI_DEV_HAS_CS_GPIOS(n),                       \
+		 DT_INST_SPI_DEV_CS_GPIOS_FLAGS(n))
+#define DRV_INST_LOCAL_MAC_ADDRESS(n)                                  \
```
have an empty line before

nandojve replied:
done!

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx.h:
```diff
@@ -89,6 +89,7 @@ struct rf2xx_dt_spi_t {
 
 struct rf2xx_config {
 	u8_t inst;
+	u8_t has_mac;
```
could you move these 2 attributes at the end of the structure, for better (future) alignment ?

nandojve replied:
done!

tbursztyka commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -3,7 +3,7 @@
 
 /*
- * Copyright (c) 2019-2020 Gerson Fernando Budke
+ * Copyright (c) 2019-2020 Gerson Fernando Budke <nandojve@gmail.com>
```
in your commit message: "does not allow user to define", "user has"

tbursztyka commented on include/drivers/ieee802154/rf2xx.h:
```diff
@@ -132,4 +132,17 @@ struct rf2xx_context {
+ *
+ * @return pointer to mac_addr variable defined on the driver context
+ */
+u8_t *rf2xx_get_mac(struct device *dev);
```
Nack on that one. 
First because it is not generic among all 15.4 drivers. But also because it is useless: you just need to set the extended address to the value you want and that's it and there is already a generic way to set it by the user.

nandojve replied:
I understood the reason but I can't find a good solution. The issue #23193 explain that MAC must be set "before" the driver init. This is because RF2xx is a external (not built in on SoC) radio. On real application the EUI-64 will be read from some memory and applied on the transceiver. This step is only possible when app already started. For instance, OpenThread si already started when user can run NET_REQUEST_IEEE802154_SET_EXT_ADDR. This could let driver send packages with different MACs. I'm assume this behavior could block a certification process, not sure about it. Kconfig and DT won't work here and only are valid for tests purposes.

That said, I agreed this is not generic among all 15.4 drivers. I agree in part on the second argument about be useless, it's a diffente way to get same goal :D. Anyway, I was wondering if driver could be blocked until it receives some API commands like NET_REQUEST_IEEE802154_SET_EXT_ADDR (and others maybe). This way, driver can be enabled but will be fully operational only when receives such commands.. It could not be like others but may should solve the problem without harm to much.

Is acceptable block driver that way? Any options about it? Implications on OT and/or Zigbee?

jukkar replied:
It looks like there is a simple way to prevent the driver from sending packets with wrong MAC. Just call `net_if_flag_set(iface, NET_IF_NO_AUTO_START);` at the end of `rf2xx_iface_init()`, that will prevent the network interface from coming up automatically. Then just call `net_if_up()` after receiving `NET_REQUEST_IEEE802154_SET_EXT_ADDR`.

tbursztyka reviewed (changes requested):
no need of the 2nd patch

nandojve committed:
drivers: ieee802154: rf2xx: Add no auto start option

Add a Kconfig option that allows user to set any necessary config
using management interface before interface be operational. A use
case is set the EUI-64 address from an external EEPROM by the
NET_REQUEST_IEEE802154_SET_EXT_ADDR command. After all configs
are done net_if_up() can be invoked to bring interface up.

Fixes #23193.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
The new proposal: if for any reason _CONFIG_IEEE802154_RF2XX_NET_IF_NO_AUTO_START_ should be defined as _true_, who did that should call _net_if_up()_ at some point.

jukkar commented on drivers/ieee802154/Kconfig.rf2xx:
```diff
@@ -18,6 +18,19 @@ config IEEE802154_RF2XX_DRV_NAME
 
+config IEEE802154_RF2XX_NET_IF_NO_AUTO_START
+	bool "RF2X must wait configuration before change to operational mode"
+	default n
```
Setting the value to 'n' is not needed as it is the default value. So this does no harm but there is a general guideline not to add these the Kconfig files.

nandojve replied:
I'll fix this, tks!

nandojve replied:
Done!

jukkar reviewed (commented):
Looks good, one nit about the config option.
